Applications/TransferLearning/Draw_Classification/jni/main.cpp

```diff
@@ -39,184 +39,201 @@
+    delete in;
+    throw std::runtime_error("Input size does not match the required size");
   }
 
-  if (interpreter->Invoke() != kTfLiteOk) {
-    std::cout << "Failed to invoke!" << std::endl;
-    exit(0);
+  int input_idx = tflite_data.interpreter->inputs()[0];
+  for (int l = 0; l < tflite_data.output_number_of_pixels; l++) {
+    (tflite_data.interpreter->typed_tensor<float>(input_idx))[l] =
+      ((float)in[l] - 127.5f) / 127.5f;
   }
 
-  output = interpreter->typed_output_tensor<float>(0);
+  if (tflite_data.interpreter->Invoke() != kTfLiteOk)
+    std::runtime_error("Failed to invoke.");
 
-  for (int l = 0; l < 128; l++) {
+  float *output = tflite_data.interpreter->typed_output_tensor<float>(0);
+  for (int l = 0; l < INPUT_SIZE; l++) {
     feature_input[l] = output[l];
   }
 
-  delete[] input_idx_list;
-  delete[] output_idx_list;
   delete[] in;
 }
 
 /**
- * @brief     Extract the features from all three categories
- * @param[in] p data path
- * @param[out] feature_input save output of tflite
- * @param[out] feature_output save label data
+ * @brief     Extract the features from pretrained model
+ * @param[in] data_path data path
+ * @param[out] input_data output of tflite model (input for the nntrainer model)
+ * @param[out] label_data one hot label data
  */
-void ExtractFeatures(std::string p, vector<vector<float>> &feature_input,
-                     vector<vector<float>> &feature_output) {
-  string total_label[TOTAL_LABEL_SIZE] = {"happy", "sad", "soso"};
-
-  int trainingSize = TOTAL_LABEL_SIZE * TOTAL_DATA_SIZE;
-
-  feature_input.resize(trainingSize);
-  feature_output.resize(trainingSize);
-
-  int count = 0;
-
-  for (int i = 0; i < TOTAL_LABEL_SIZE; i++) {
-    std::string path = p;
-    path += total_label[i];
+void extractFeatures(const TFLiteData &tflite_data,
+                     std::vector<std::vector<float>> &input_data,
+                     std::vector<std::vector<float>> &label_data) {
+  int trainingSize = LABEL_SIZE * NUM_DATA_PER_LABEL;
+
+  input_data.resize(trainingSize, std::vector<float>(INPUT_SIZE));
+  /** resize label data to size and initialize to 0 */
+  label_data.resize(trainingSize, std::vector<float>(LABEL_SIZE, 0));
+
+  for (int i = 0; i < LABEL_SIZE; i++) {
+    for (int j = 0; j < NUM_DATA_PER_LABEL; j++) {
+      std::string label_file = label_names[i] + std::to_string(j + 1) + ".bmp";
+      std::string img =
+        tflite_data.data_path + "/" + label_names[i] + "/" + label_file;
+
+      int count = i * NUM_DATA_PER_LABEL + j;
+      getInputFeature(tflite_data, img, input_data[count]);
+      label_data[count][i] = 1;
+    }
+  }
+}
 
-    for (int j = 0; j < TOTAL_DATA_SIZE; j++) {
-      std::string img = path + "/";
-      img += total_label[i] + std::to_string(j + 1) + ".bmp";
-      printf("%s\n", img.c_str());
+/**
+ * Data generator callback
+ */
+int getBatch_train(float **input, float **label, bool *last, void *user_data) {
+  if (iteration >= EPOCH_SIZE) {
```

Comment:
How about `static int iteration = 0` instead of the global variable?

Reply:
done.